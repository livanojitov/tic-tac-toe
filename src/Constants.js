const EMPTY = 0;
const COMPUTER = 1;
const USER = 2;
const TOTAL_SQUARES = 9;
const RANDOM = 1;
const HARD = 2;
const LOST = 0;
const WON = 1;
const DRAW = 2;
const UNKNOWN = 3;
const ENGLISH = 0;
const FRENCH = 1;
const SPANISH = 2;
const IMAGES_PER_CATEGORY = 20;

export {
  EMPTY,
  COMPUTER,
  USER,
  RANDOM,
  HARD,
  TOTAL_SQUARES,
  LOST,
  WON,
  DRAW,
  UNKNOWN,
  ENGLISH,
  FRENCH,
  SPANISH,
  IMAGES_PER_CATEGORY,
};
