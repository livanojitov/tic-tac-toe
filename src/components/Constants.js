const EMPTY = 0;

const COMPUTER = 1;
const USER = 2;

const EASY = 1;
const HARD = 2;
const NORMAL = 3;

const TOTAL_SQUARES = 9;

const CATEGORY = 'Category';

const YOU = 'You';
const OPPONENT = 'Computer';
const REFRESH = 'Refresh';
const USER_IMAGE_TOOLTIP = 'You will play with this image';
const COMPUTER_IMAGE_TOOLTIP = 'The computer will play with this image';
const REFRESH_TOOLTIP = 'Click to refresh the two images';

const START = 'Start';
const YES = 'Yes';
const NO = 'No';
const START_TOOLTIP = 'Who starts playing the game: You or the computer?';
const YES_TOOLTIP = 'You start playing the game.';
const NO_TOOLTIP = 'The computer starts playing the game.';
const GAME_STARTED_BY = 'Game started by:';

const LEVEL = 'Level';
const COMPUTER_PLAYS_EASY = 'Easy';
const COMPUTER_PLAYS_HARD = 'Hard';
const COMPUTER_PLAYS_NORMAL = 'Normal';
const LEVEL_TOOLTIP = 'How much thinking the computer does before playing.';
const COMPUTER_PLAYS_EASY_TOOLTIP = 'The computer plays randomly.';
const COMPUTER_PLAYS_HARD_TOOLTIP = 'The computer does some thinking before playing.';
const COMPUTER_PLAYS_NORMAL_TOOLTIP = 'The computer sometimes plays easy and sometimes plays hard.';
const LEVEL_HISTORY_EASY = 'Level: Easy';
const LEVEL_HISTORY_HARD = 'Level: Hard';
const LEVEL_HISTORY_NORMAL = 'Level: Normal';

const PLAY = 'Start the game';
const GAME_OVER = 'Game over';
const PLAY_AGAIN = 'Play again';
const LOST = 'You lost!';
const WON = 'You won!';
const DRAW = 'It\'s a draw!';

const GAME_WRITTEN_WITH = 'This game is written with:';
const SOURCE_CODE = 'Source Code:';
const OTHER_GAMES = 'Other games:';

const AUTHOR = 'Author:';
const EMAIL = 'Email:';
const GITHUB = 'Github:';
const LINKEDIN = 'LinkedIn:';
const HIRE_ME = 'Hire me!';

const NO_HISTORY_YET = 'No history yet. Play some games and comeback.';

const HOME = 'Home';
const HISTORY = 'History';
const ABOUT = 'About';
const CONTACT = 'Contact';

export { EMPTY, COMPUTER, USER, EASY, HARD, NORMAL, TOTAL_SQUARES, CATEGORY, YOU, OPPONENT, REFRESH, USER_IMAGE_TOOLTIP, COMPUTER_IMAGE_TOOLTIP, REFRESH_TOOLTIP, START, YES, NO, START_TOOLTIP, YES_TOOLTIP, NO_TOOLTIP, GAME_STARTED_BY, LEVEL, COMPUTER_PLAYS_EASY, COMPUTER_PLAYS_HARD, LEVEL_TOOLTIP, COMPUTER_PLAYS_EASY_TOOLTIP, COMPUTER_PLAYS_HARD_TOOLTIP, LEVEL_HISTORY_EASY, LEVEL_HISTORY_HARD, PLAY, GAME_OVER, PLAY_AGAIN, LOST, WON, DRAW, GAME_WRITTEN_WITH, SOURCE_CODE, OTHER_GAMES, AUTHOR, EMAIL, GITHUB, LINKEDIN, HIRE_ME, NO_HISTORY_YET, HOME, HISTORY, ABOUT, CONTACT, COMPUTER_PLAYS_NORMAL, COMPUTER_PLAYS_NORMAL_TOOLTIP, LEVEL_HISTORY_NORMAL };  
