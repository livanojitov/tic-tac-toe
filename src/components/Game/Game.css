.game{
  width: 100%;
  margin: 10px auto;
}

.play-again{
  margin-left: 5px;
}

.start-playing{
  margin-bottom: 10px;
}