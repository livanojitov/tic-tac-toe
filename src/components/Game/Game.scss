.game {
  width: 100%;
  margin: 10px auto;
}

.play-again {
  margin-left: 5px;
}

.start-playing {
  margin-bottom: 10px;
}

.buttons {
  margin-top: 5px;

  .see-game {
    margin: 0 5px;

    a {
      color: black;
    }
  }

  .delete-game {
    margin: 0 5px;
    cursor: pointer;
  }

  .back {
    cursor: pointer;
  }
}

hr {
  margin: 15px;
}
