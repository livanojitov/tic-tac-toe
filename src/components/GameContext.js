import React, { useState, createContext, useEffect } from 'react';
export const GameContext = createContext();

const GameContextProvider = (props) => {

  const [games, setGames] = useState([]);

  useEffect(() => {
    if (typeof(Storage) !== "undefined") {
      if (games.length){
        localStorage.tictactoe = JSON.stringify(games);
      }else{
        if (GameContextProvider.delete === 1){
          GameContextProvider.delete = 0;
          localStorage.removeItem('tictactoe');
        }else{
          if (localStorage.tictactoe){
            setGames(JSON.parse(localStorage.tictactoe));
          }
        }
      }
    }
  },[games]);

  const addGame = (game) => {
    game.id = Math.floor(Math.random() * 10000);
    setGames([...games, game]);
  };

  const removeGame = (id) => {
    if (games.length > 0){
      GameContextProvider.delete = 1;
      id = id * 1;
      setGames(games.filter(game => game.id !== id));   
    } 
  }

  const getGame = (id) => {
    if (games.length > 0){
      id = id * 1;
      return games.filter((game) => game.id === id)[0];
    }else{
      return false;
    }    
  }

  return (
    <GameContext.Provider value={{ games, addGame, removeGame, getGame }}>
      {props.children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;