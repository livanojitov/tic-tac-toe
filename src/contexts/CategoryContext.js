import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";

export const CategoryContext = createContext();

const CategoryContextProvider = ({ children }) => {
  let { data: categories, error } = useFetchData("categories");

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  return (
    <CategoryContext.Provider value={{ categories }}>
      {children}
    </CategoryContext.Provider>
  );
};

export default CategoryContextProvider;
