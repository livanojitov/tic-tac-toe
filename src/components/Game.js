import React           from 'react';
import Category        from './Category';
import StartGame       from './StartGame';
import BoardUI         from './BoardUI';
import Level           from './Level';
import TicTacToe       from './TicTacToe';
import Info            from './Info';
import { GameContext } from './GameContext';

class Game extends React.Component {
 
  computer = 1;
  user = 2;
  static contextType = GameContext;

  constructor(props){
    super(props);
    this.startingPlayer = -1;
    this.player = -1;
    this.level  = -1;
    this.timeout = 1;
    this.game = new TicTacToe();
    this.message = '';
    this.state = {
      category: {},
      disableBoard: true,
      reset: false,
      board: this.game.getBoard(),
      winningSquares: [],
      gameOver: false,
      showStartButton: true
     };
  }

  render(){
    return(
      <div className="game">
        <Category onCategoryChange = {this.setCategory}/>  
        <div className="questions">
          <StartGame disable={!this.state.disableBoard} onPlayerChange = {this.setStartingPlayer} />
          <Level     disable={!this.state.disableBoard} onLevelChange  = {this.setLevel } />
        </div>
        {this.state.showStartButton &&  ( <div className="start-playing"><button onClick={this.gameInit}>Start the game</button></div>)}
        {(this.state.category.imageUser) && (
          <BoardUI category       = {this.state.category} 
                   disableBoard   = {this.state.disableBoard}
                   board          = {this.state.board}
                   onPlayUser     = {this.gameStarts}
                   winningSquares = {this.state.winningSquares}
                   gameOver       = {this.state.gameOver} />)}
        {this.state.gameOver && ( <Info message = {this.message} startOver = {this.startOver}/>)}                  
        </div>
    )
  }

  gameInit = () => {
    this.setState(() => ({disableBoard: false, showStartButton: false}));
    this.game.setLevel(this.level);
    this.game.setPlayer(this.startingPlayer);
    this.gameStarts();
  }

  gameStarts = (e) => {
    if (!this.state.gameOver){
      if (this.player === this.computer){
        if (this.game.isEmpty()){
          this.game.computer.playRandomly();
        }else{
          this.game.computer.play(); 
        }
        this.setState(() => ({board : this.game.getBoard()}));
        let winnerSquares = this.game.isAWinner(this.computer);
        if (winnerSquares){
          this.gameOver("You lost!", winnerSquares);
        }else if (this.game.isFull()){
          this.gameOver("It's a draw!");
        }else{
          this.player = this.user;
          window.setTimeout(this.gameStarts, this.timeout);
        }  
      }else if (this.player === this.user){
        if (e){
          this.game.user.play(e.target.id);      
          this.setState(() => ({board : this.game.getBoard()}));
          let winnerSquares = this.game.isAWinner(this.user);
          if (winnerSquares){
            this.gameOver("You won!", winnerSquares);
          }else if (this.game.isFull()){
            this.gameOver("It's a draw!");
          }else{
            this.player = this.computer;
            window.setTimeout(this.gameStarts, this.timeout);
          } 
        }else{
          window.setTimeout(this.gameStarts, this.timeout);
        } 
      }
    }  
  }
  
  startOver = () => {
    this.message = '';
    this.game.reset();
    this.setState(() => {
        return {
          winningSquares : [],
          gameOver : false,
          board: this.game.getBoard(), 
          disableBoard: false,
          showStartButton: false
        }
    }, () => {
      this.game.setLevel(this.level);
      this.game.setPlayer(this.startingPlayer);
      this.player = this.startingPlayer;
      this.gameStarts();
    });
  }

  gameOver = (message, winningSquares = []) => {
    this.message = message;
    if (winningSquares.length){
      const ws = this.state.winningSquares;
      [ws[0], ws[1], ws[2]] = winningSquares;
      this.setState(() => ({ gameOver : true, disableBoard: true, showStartButton: false, winningSquares : ws }));
    }else{
      this.setState(() => ({ gameOver : true, disableBoard: true }));
    } 
    const { addGame } = this.context;
    addGame({
      board          : [...this.game.getBoard()],
      winningSquares : [...winningSquares],
      message,
      whoStarted     : this.startingPlayer,
      level          : this.level,
      ...this.state.category,
    });
  }

  setCategory = (category) => {
    this.setState(() => ({category}));
  }

  setStartingPlayer = (player) => {
    this.startingPlayer = player;
    this.player = player;
  }

  setLevel = (level) => {
    this.level = level;
  }

}

export default Game