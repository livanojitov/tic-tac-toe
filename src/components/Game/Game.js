import React, { useContext } from "react";
import { GameDBContext } from "../../contexts/GameDBContext";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameContext } from "../../contexts/GameContext";
import Category from "../Category/Category";
import Images from "../Images/Images";
import StartGame from "../StartGame/StartGame";
import Level from "../Level/Level";
import BoardUI from "../Board/BoardUI";
import * as constants from "../../Constants";
import useGame from "../../hooks/useGame";
import "./Game.scss";

const { LOST, WON } = constants;

const Game = () => {
  const { dispatch } = useContext(GameDBContext);
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch: dispatchGame } = useContext(GameContext);

  const { gameInitHandler, gameOverHandler, result } = useGame(
    dispatch,
    dispatchGame
  );

  return dictionary ? (
    <div className="game text-center" data-testid="game">
      <Category />

      <Images />

      <div className="settings">
        <StartGame />
        <Level />
      </div>

      <BoardUI onGameOver={gameOverHandler} />
      <div className="board-footer">
        {game.over && (
          <>
            <span>{dictionary.GAME_OVER}: </span>
            <span data-testid="message">
              {result === LOST
                ? dictionary.MESSAGE_LOST
                : result === WON
                ? dictionary.MESSAGE_WON
                : dictionary.MESSAGE_DRAW}
            </span>
          </>
        )}

        {((!result && game.disabled) || game.over) && (
          <button
            data-testid="play"
            className="play-again"
            onClick={gameInitHandler}
          >
            {game.over ? dictionary.PLAY_AGAIN : dictionary.PLAY}
          </button>
        )}
      </div>
    </div>
  ) : (
    <div className="loading" data-testid="loading">
      Loading the game section of the page...
    </div>
  );
};

export default Game;
