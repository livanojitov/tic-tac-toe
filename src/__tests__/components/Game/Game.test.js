import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { CategoryContext } from "../../../contexts/CategoryContext";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import { GameDBContext } from "../../../contexts/GameDBContext";
import { GameContext } from "../../../contexts/GameContext";
import Game from "../../../components/Game/Game";
import useGame from "../../../hooks/useGame";
import { USER, RANDOM, LOST, WON, DRAW } from "../../../Constants";

jest.mock("../../../components/Board/BoardUI", () => () => (
  <label>Board</label>
));

jest.mock("../../../hooks/useGame");

describe("Game", () => {
  let dictionary;
  let categories;
  let game;
  let dispatch;
  let error;

  beforeEach(() => {
    dictionary = {
      ANIMALS: "Animals",
      FRUITS_AND_VEGETABLES: "Fruits & Vegetables",
      GERMANY_LANDMARKS: "Germany landmarks",
      HAVANA_LANDMARKS: "Havana landmarks",
      MONTREAL_LANDMARKS: "Montreal landmarks",
      PUPPIES: "Puppies",
      SEINFELD: "Seinfeld",
      SOCCER_PLAYERS: "Soccer players",
      SPAIN_LANDMARKS: "Spain landmarks",
      VANCOUVER_LANDMARKS: "Vancouver landmarks",
      CATEGORY: "Category",
      PLAY: "Play",
      PLAY_AGAIN: "Play again",
      MESSAGE_WON: "You won!",
      MESSAGE_LOST: "You lost!",
      MESSAGE_DRAW: "It's a draw!",
    };

    categories = [
      { id: 0, name: "HAVANA_LANDMARKS", folder: "habana" },
      { id: 1, name: "MONTREAL_LANDMARKS", folder: "montreal" },
      { id: 2, name: "VANCOUVER_LANDMARKS", folder: "vancouver" },
      { id: 3, name: "SPAIN_LANDMARKS", folder: "spain" },
      { id: 4, name: "GERMANY_LANDMARKS", folder: "germany" },
      { id: 5, name: "SOCCER_PLAYERS", folder: "soccer" },
      { id: 6, name: "FRUITS_AND_VEGETABLES", folder: "fruits" },
      { id: 7, name: "ANIMALS", folder: "animals" },
      { id: 8, name: "PUPPIES", folder: "puppies" },
      { id: 9, name: "SEINFELD", folder: "seinfeld" },
    ];

    game = {
      category: {
        id: 0,
        folder: "habana",
        name: "HAVANA_LANDMARKS",
      },
      images: {
        user: 0,
        computer: 1,
      },
      first: USER,
      level: RANDOM,
      disabled: true,
      over: false,
    };

    dispatch = jest.fn();

    error = "";
  });

  const renderGame = (result = "") => {
    useGame.mockReturnValue({
      gameInitHandler: jest.fn,
      gameOverHandler: jest.fn,
      result,
    });

    return render(
      <DictionaryContext.Provider value={{ dictionary, error }}>
        <GameDBContext.Provider value={{ dispatch }}>
          <CategoryContext.Provider value={{ categories }}>
            <GameContext.Provider value={{ game, dispatch }}>
              <Game />
            </GameContext.Provider>
          </CategoryContext.Provider>
        </GameDBContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  it("renders correctly", () => {
    const { asFragment } = renderGame();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not display the Game info if dictionary hasn't been fetched yet", () => {
    dictionary = "";

    renderGame();

    expect(screen.queryByTestId("game")).toBeNull();
  });

  it("should display a loading message while fetching the dictionary", () => {
    dictionary = "";

    renderGame();

    expect(screen.getByTestId("loading")).not.toBeNull();
  });

  it("should display the game lost message", () => {
    game.over = true;

    renderGame(LOST);

    expect(screen.getByTestId("message")).toHaveTextContent(
      dictionary.MESSAGE_LOST
    );
  });

  it("should display the game won message", () => {
    game.over = true;

    renderGame(DRAW);

    expect(screen.getByTestId("message")).toHaveTextContent(
      dictionary.MESSAGE_DRAW
    );
  });

  it("should display the game draw message", () => {
    game.over = true;

    renderGame(WON);

    expect(screen.getByTestId("message")).toHaveTextContent(
      dictionary.MESSAGE_WON
    );
  });
});
