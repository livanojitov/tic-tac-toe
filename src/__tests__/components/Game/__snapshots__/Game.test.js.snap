// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<DocumentFragment>
  <div
    class="game text-center"
    data-testid="game"
  >
    <div
      class="categories"
      data-testid="categories"
    >
      <span
        data-testid="description"
      >
        Category: 
      </span>
      <select
        data-testid="select"
      >
        <option
          data-testid="select-option"
          value="0"
        >
          Havana landmarks
        </option>
        <option
          data-testid="select-option"
          value="1"
        >
          Montreal landmarks
        </option>
        <option
          data-testid="select-option"
          value="2"
        >
          Vancouver landmarks
        </option>
        <option
          data-testid="select-option"
          value="3"
        >
          Spain landmarks
        </option>
        <option
          data-testid="select-option"
          value="4"
        >
          Germany landmarks
        </option>
        <option
          data-testid="select-option"
          value="5"
        >
          Soccer players
        </option>
        <option
          data-testid="select-option"
          value="6"
        >
          Fruits & Vegetables
        </option>
        <option
          data-testid="select-option"
          value="7"
        >
          Animals
        </option>
        <option
          data-testid="select-option"
          value="8"
        >
          Puppies
        </option>
        <option
          data-testid="select-option"
          value="9"
        >
          Seinfeld
        </option>
      </select>
    </div>
    <div
      class="images"
      data-testid="images"
    >
      <div
        class="user-image"
      >
        <span>
          : 
        </span>
        <img
          alt="user"
          class="image-token"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F0.jpg?alt=media"
        />
        <img
          alt="refresh"
          class="refresh"
          data-testid="user-image"
          id="user"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Frefresh.png?alt=media"
        />
      </div>
      <div
        class="computer-image"
      >
        <span>
          : 
        </span>
        <img
          alt="computer"
          class="image-token"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fhabana%2F1.jpg?alt=media"
        />
        <img
          alt="refresh"
          class="refresh"
          data-testid="computer-image"
          id="computer"
          src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Frefresh.png?alt=media"
        />
      </div>
    </div>
    <div
      class="settings"
    >
      <div
        class="start"
        data-testid="startgame"
      >
        <span>
          ?
        </span>
        <select
          data-testid="select"
        >
          <option
            data-testid="select-option"
            selected=""
            value="2"
          />
          <option
            data-testid="select-option"
            value="1"
          />
        </select>
      </div>
      <div
        class="level"
        data-testid="level"
      >
        <span>
          ?
        </span>
        <select
          data-testid="select"
        >
          <option
            data-testid="select-option"
            selected=""
            value="1"
          >
             
          </option>
          <option
            data-testid="select-option"
            value="2"
          >
             
          </option>
        </select>
      </div>
    </div>
    <label>
      Board
    </label>
    <div
      class="board-footer"
    >
      <button
        class="play-again"
        data-testid="play"
      >
        Play
      </button>
    </div>
  </div>
</DocumentFragment>
`;
